### PR TITLE
[Feature][Task Plugin] Add Dynamic task to generate dynamic worklfows with list parameters

### LINKED ISSUE #14125: [Feature][Task Plugin] Add Dynamic task to generate dynamic worklfows with list parameters
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

DOC: https://github.com/jieguangzhou/dolphinscheduler/blob/feat/task-plugin-dynamic/docs/docs/en/guide/task/dynamic.md

image

image

We can get the output with `${dynamic.out(taskName)}` like this

```json
[
    {
        "dynParams":{
            "a":"a1",
            "b":"b1"
        },
        "outputValue":{
            "p":"a1-b1"
        },
        "mappedTimes":1
    },
    {
        "dynParams":{
            "a":"a2",
            "b":"b1"
        },
        "outputValue":{
            "p":"a2-b1"
        },
        "mappedTimes":2
    },
    {
        "dynParams":{
            "a":"a3",
            "b":"b1"
        },
        "outputValue":{
            "p":"a3-b1"
        },
        "mappedTimes":3
    }
]

```

image

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by jieguangzhou**
## Purpose of the pull request

close: #14125

DOC: https://github.com/jieguangzhou/dolphinscheduler/blob/feat/task-plugin-dynamic/docs/docs/en/guide/task/dynamic.md

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

**Comment by jieguangzhou** [comment]
I will add UT later, please help to review the base codes

File: docs/docs/en/guide/task/dynamic.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+|        **Task Parameters**        |                                                                                                                         **Description**                                                                                                                          |
+|-----------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Child Node                        | Select the workflow definition of the sub-workflow. You can jump to the workflow definition of the selected sub-workflow by entering the sub-node in the upper right corner.                                                                                     |
+| max num of sub workflow isntances | The maximum number of sub-workflow instances dynamically generated. After exceeding this upper limit, the dynamically generated sub-workflow instances will no longer be executed.                                                                               |
```
Comment (Radeity) [inline docs/docs/en/guide/task/dynamic.md]: isntances -> instances

Comment (jieguangzhou) [reply docs/docs/en/guide/task/dynamic.md]: Thanks, I have fixed that

File: dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_h2.sql [hunk 2]
```diff
@@ -2042,3 +2042,16 @@ CREATE TABLE t_ds_trigger_relation
...
+CREATE TABLE t_ds_relation_sub_workflow (
+    id BIGINT AUTO_INCREMENT NOT NULL,
+    parent_workflow_instance_id BIGINT NOT NULL,
+    parent_task_code BIGINT NOT NULL,
+    sub_workflow_instance_id BIGINT NOT NULL,
+    PRIMARY KEY (id),
+    INDEX idx_parent_workflow_instance_id (parent_workflow_instance_id),
+    INDEX idx_parent_task_code (parent_task_code),
+    INDEX idx_sub_workflow_instance_id (sub_workflow_instance_id)
+) ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8;
```
Comment (zhongjiajie) [inline dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_h2.sql:2057]:
```suggestion
);
```

**Review by zhongjiajie — COMMENTED** [review]
Others LGTM

File: dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_h2.sql (hunk 2 above)
Comment (jieguangzhou) [reply dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_h2.sql:2057]: Done

File: dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/mapper/RelationSubWorkflowMapper.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import com.baomidou.mybatisplus.core.mapper.BaseMapper;
+
+/**
+ * process instance map mapper interface
+ */
+public interface RelationSubWorkflowMapper extends BaseMapper<RelationSubWorkflow> {
+
+    int batchInsert(@Param("relationSubWorkflows") List<RelationSubWorkflow> relationSubWorkflows);
+
+    List<RelationSubWorkflow> selectAllSubProcessInstance(@Param("parentWorkflowInstanceId") Long parentWorkflowInstanceId,
```
Comment (caishunfeng) [inline dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/mapper/RelationSubWorkflowMapper.java]:
```suggestion
    List<RelationSubWorkflow> queryAllSubProcessInstance(@Param("parentWorkflowInstanceId") Long parentWorkflowInstanceId,
```

File: dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/repository/ProcessInstanceDao.java [hunk 4]
```diff
@@ -38,6 +38,8 @@ public interface ProcessInstanceDao {
      */
     public int upsertProcessInstance(ProcessInstance processInstance);
 
+    List<ProcessInstance> selectBatchIds(List<Long> processInstanceIds);
```
Comment (caishunfeng) [inline dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/repository/ProcessInstanceDao.java]:
```suggestion
    List<ProcessInstance> queryBatchIds(List<Long> processInstanceIds);
```

File: dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/RelationSubWorkflowMapper.xml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        id, parent_workflow_instance_id, parent_task_code, sub_workflow_instance_id
+    </sql>
+    <insert id="batchInsert">
+        insert into t_ds_relation_sub_workflow (parent_workflow_instance_id, parent_task_code, sub_workflow_instance_id)
+        values
+        <foreach collection="relationSubWorkflows" item="relationSubWorkflow" separator=",">
+            (#{relationSubWorkflow.parentWorkflowInstanceId}, #{relationSubWorkflow.parentTaskCode}, #{relationSubWorkflow.subWorkflowInstanceId})
+        </foreach>
+    </insert>
+    <select id="selectAllSubProcessInstance" resultType="org.apache.dolphinscheduler.dao.entity.RelationSubWorkflow">
```
Comment (caishunfeng) [inline dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/RelationSubWorkflowMapper.xml]:
```suggestion
    <select id="queryAllSubProcessInstance" resultType="org.apache.dolphinscheduler.dao.entity.RelationSubWorkflow">
```

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/dynamic/DynamicLogicTask.java [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        }
+        return dynamicInputParameters;
+    }
+
+    @Override
+    public void kill() {
+        try {
+            changeRunningSubprocessInstancesToStop(WorkflowExecutionStatus.READY_STOP);
+        } catch (MasterTaskExecuteException e) {
+            log.error("kill " + taskInstance.getName() + " error", e);
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/dynamic/DynamicLogicTask.java]:
```suggestion
            log.error("kill {} error", taskInstance.getName(), e);
```

**Review by caishunfeng — DISMISSED** [review]
LGTM overall

File: dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/mapper/RelationSubWorkflowMapper.java (hunk 3 above)
Comment (jieguangzhou) [reply dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/mapper/RelationSubWorkflowMapper.java]: done

File: dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/repository/ProcessInstanceDao.java (hunk 4 above)
Comment (jieguangzhou) [reply dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/repository/ProcessInstanceDao.java]: done

**Review by devosend — APPROVED** [review]
LGTM

**Pull request merged**